File: backend/pipeline/retainpdf_pipeline/ocr/ocr_provider/paddle_markdown.py
from __future__ import annotations

import base64
import re
from pathlib import Path

from retainpdf_pipeline.ocr.retry import RetainNetworkError
from retainpdf_pipeline.ocr.retry import direct_session
from retainpdf_pipeline.ocr.retry import request_with_retry

_MARKDOWN_IMAGE_SESSION = direct_session(pool_connections=4, pool_maxsize=4)
_PAGE_PREFIX_RE = re.compile(r"^page-\d+(/|$)")
_IMG_SRC_RE = re.compile(r"""(<img\b[^>]*\bsrc=["'])([^"']+)(["'])""", re.IGNORECASE)
_MARKDOWN_IMG_SRC_RE = re.compile(
    r"""(!\[[^\]\n]*\]\(\s*<?)([^\s)>]+)(>?\s*(?:["'][^"']*["']\s*)?\))"""
)
_IMG_TAG_RE = re.compile(r"""<img\b([^>]*)>""", re.IGNORECASE)
_CENTERED_IMG_DIV_RE = re.compile(
    r"""<div\s+style=["']text-align:\s*center;?["']\s*>\s*(<img\b[^>]*>)\s*</div>""",
    re.IGNORECASE,
)
_HTML_ATTR_RE = re.compile(r"""([A-Za-z_:][-A-Za-z0-9_:.]*)\s*=\s*["']([^"']*)["']""")


def job_markdown_dir(job_root: Path) -> Path:
    return job_root / "md"


def job_markdown_images_dir(job_root: Path) -> Path:
    return job_markdown_dir(job_root) / "images"


def paddle_markdown_target_rel_path(rel_path: str, *, page_index: int) -> Path:
    normalized = str(rel_path or "").strip().lstrip("/").replace("\\", "/")
    if not normalized:
        return Path()
    if _PAGE_PREFIX_RE.match(normalized):
        return Path(normalized)
    return Path(f"page-{page_index}") / normalized


def paddle_markdown_rel_src_path(src: str, *, page_index: int) -> str:
    normalized = str(src or "").strip().replace("\\", "/")
    if not normalized or normalized.startswith(("http://", "https://", "data:")):
        return normalized
    normalized = normalized.lstrip("/")
    if normalized.startswith("images/"):
        return normalized
    return (Path("images") / paddle_markdown_target_rel_path(normalized, page_index=page_index)).as_posix()


def rewrite_paddle_markdown_image_srcs(text: str, *, page_index: int) -> str:
    rewritten = _IMG_SRC_RE.sub(
        lambda match: f"{match.group(1)}{paddle_markdown_rel_src_path(match.group(2), page_index=page_index)}{match.group(3)}",
        text,
    )
    return _MARKDOWN_IMG_SRC_RE.sub(
        lambda match: f"{match.group(1)}{paddle_markdown_rel_src_path(match.group(2), page_index=page_index)}{match.group(3)}",
        rewritten,
    )


def normalize_paddle_markdown_images(text: str, *, page_index: int) -> str:
    rewritten = rewrite_paddle_markdown_image_srcs(text, page_index=page_index)
    rewritten = _CENTERED_IMG_DIV_RE.sub(lambda match: _markdown_image_from_img_tag(match.group(1)), rewritten)
    return _IMG_TAG_RE.sub(lambda match: _markdown_image_from_attrs(match.group(1)), rewritten)


def _markdown_image_from_img_tag(img_tag: str) -> str:
    match = _IMG_TAG_RE.search(img_tag)
    if not match:
        return img_tag
    return _markdown_image_from_attrs(match.group(1))


def _markdown_image_from_attrs(attrs_text: str) -> str:
    attrs = {key.lower(): value for key, value in _HTML_ATTR_RE.findall(attrs_text)}
    src = attrs.get("src", "").strip()
    if not src:
        return f"<img{attrs_text}>"
    alt = attrs.get("alt", "Image").strip() or "Image"
    alt = alt.replace("[", "\\[").replace("]", "\\]")
    return f"![{alt}]({src})"


def decode_paddle_markdown_image(payload: str) -> bytes:
    value = str(payload or "").strip()
    if not value:
        raise RuntimeError("empty markdown image payload")
    if value.startswith(("http://", "https://")):
        try:
            response = request_with_retry(
                _MARKDOWN_IMAGE_SESSION,
                "get",
                value,
                timeout=300,
                attempts=3,
                backoff_seconds=0.5,
                label="Paddle markdown image",
            )
        except RetainNetworkError as err:
            raise RuntimeError(f"download Paddle markdown image failed: {value}: {err}") from err
        return response.content
    if value.startswith("data:") and "," in value:
        _, encoded = value.split(",", 1)
        return base64.b64decode(encoded)
    return base64.b64decode(value)


def materialize_paddle_markdown_artifacts(*, payload: dict, job_root: Path) -> Path | None:
    layout_results = payload.get("layoutParsingResults") or []
    if not isinstance(layout_results, list) or not layout_results:
        return None

    markdown_dir = job_markdown_dir(job_root)
    images_root = job_markdown_images_dir(job_root)
    page_texts: list[str] = []
    wrote_anything = False

    for page_index, page_payload in enumerate(layout_results, start=1):
        if not isinstance(page_payload, dict):
            continue
        markdown = page_payload.get("markdown") or {}
        if not isinstance(markdown, dict):
            continue
        text = str(markdown.get("text") or "")
        images = markdown.get("images") or {}
        if not text.strip() and not images:
            continue

        remapped_text = text
        if isinstance(images, dict) and images:
            for raw_rel_path, raw_image_payload in images.items():
                rel_path = str(raw_rel_path or "").strip().lstrip("/")
                if not rel_path:
                    continue
                target_rel_path = _target_rel_path_for_image_key(rel_path, remapped_text, page_index=page_index)
                if not target_rel_path.as_posix():
                    continue
                markdown_rel_path = Path("images") / target_rel_path
                target_path = images_root / target_rel_path
                target_path.parent.mkdir(parents=True, exist_ok=True)
                target_path.write_bytes(decode_paddle_markdown_image(str(raw_image_payload or "")))
                normalized_source = rel_path.replace("\\", "/")
                normalized_target = markdown_rel_path.as_posix()
                if not _is_page_prefixed_alias(target_rel_path.as_posix(), normalized_source):
                    remapped_text = remapped_text.replace(normalized_source, normalized_target)
                wrote_anything = True
        remapped_text = normalize_paddle_markdown_images(remapped_text, page_index=page_index)

        if remapped_text.strip():
            page_texts.append(remapped_text.strip())
            wrote_anything = True

    if not wrote_anything:
        return None

    markdown_dir.mkdir(parents=True, exist_ok=True)
    full_md_path = markdown_dir / "full.md"
    full_md_path.write_text("\n\n".join(page_texts).strip() + "\n", encoding="utf-8")
    return full_md_path


def _target_rel_path_for_image_key(rel_path: str, markdown_text: str, *, page_index: int) -> Path:
    normalized = str(rel_path or "").strip().lstrip("/").replace("\\", "/")
    if not normalized:
        return Path()
    if _PAGE_PREFIX_RE.match(normalized):
        return Path(normalized)
    for match in _IMG_SRC_RE.finditer(markdown_text):
        src = match.group(2).strip().lstrip("/").replace("\\", "/")
        if _PAGE_PREFIX_RE.match(src) and src.endswith(normalized):
            return Path(src)
    for match in _MARKDOWN_IMG_SRC_RE.finditer(markdown_text):
        src = match.group(2).strip().lstrip("/").replace("\\", "/")
        if _PAGE_PREFIX_RE.match(src) and src.endswith(normalized):
            return Path(src)
    return paddle_markdown_target_rel_path(normalized, page_index=page_index)


def _is_page_prefixed_alias(target_rel_path: str, source_rel_path: str) -> bool:
    target = str(target_rel_path or "").replace("\\", "/").lstrip("/")
    source = str(source_rel_path or "").replace("\\", "/").lstrip("/")
    return target != source and _PAGE_PREFIX_RE.match(target) and target.endswith(source)

File: backend/packages/retain-jobs/src/job_runner/ocr_flow/paddle_markdown.rs
use std::path::{Path, PathBuf};

use anyhow::{anyhow, Context, Result};
use base64::engine::general_purpose::STANDARD;
use base64::Engine;
use regex::{Captures, Regex};
use reqwest::Client;
use serde_json::Value;

pub(super) async fn materialize_paddle_markdown_artifacts(
    payload: &Value,
    job_root: &Path,
) -> Result<Option<PathBuf>> {
    let Some(layout_results) = payload
        .get("layoutParsingResults")
        .and_then(Value::as_array)
    else {
        return Ok(None);
    };
    if layout_results.is_empty() {
        return Ok(None);
    }

    let markdown_dir = job_root.join("md");
    let images_root = markdown_dir.join("images");
    let http = Client::new();
    let mut page_texts = Vec::new();
    let mut wrote_anything = false;

    for (page_idx, page_payload) in layout_results.iter().enumerate() {
        let Some(markdown) = page_payload.get("markdown").and_then(Value::as_object) else {
            continue;
        };
        let text = markdown
            .get("text")
            .and_then(Value::as_str)
            .unwrap_or("")
            .trim()
            .to_string();
        let images = markdown.get("images").and_then(Value::as_object);
        if text.is_empty() && images.is_none() {
            continue;
        }

        let mut remapped_text = text;
        if let Some(images) = images {
            for (raw_rel_path, raw_payload) in images {
                let rel_path = raw_rel_path.trim().trim_start_matches('/');
                if rel_path.is_empty() {
                    continue;
                }
                // Preserve the provider-returned relative path shape under md/images.
                // The page prefix prevents multi-page jobs from colliding on identical image names.
                let target_rel = paddle_markdown_target_rel_path_for_image_key(
                    rel_path,
                    &remapped_text,
                    page_idx + 1,
                );
                let markdown_rel = PathBuf::from("images").join(&target_rel);
                let target_path = images_root.join(&target_rel);
                let image_bytes = decode_markdown_image_payload(&http, raw_payload).await?;
                if let Some(parent) = target_path.parent() {
                    tokio::fs::create_dir_all(parent).await?;
                }
                tokio::fs::write(&target_path, image_bytes)
                    .await
                    .with_context(|| format!("failed to write {}", target_path.display()))?;
                if !is_page_prefixed_alias(&target_rel.to_string_lossy(), rel_path) {
                    remapped_text =
                        remapped_text.replace(rel_path, &markdown_rel.to_string_lossy());
                }
                wrote_anything = true;
            }
        }
        remapped_text = normalize_paddle_markdown_images(&remapped_text, page_idx + 1);

        if !remapped_text.trim().is_empty() {
            page_texts.push(remapped_text.trim().to_string());
            wrote_anything = true;
        }
    }

    if !wrote_anything {
        return Ok(None);
    }

    tokio::fs::create_dir_all(&markdown_dir).await?;
    let full_md_path = markdown_dir.join("full.md");
    let mut content = page_texts.join("\n\n");
    if !content.ends_with('\n') {
        content.push('\n');
    }
    tokio::fs::write(&full_md_path, content)
        .await
        .with_context(|| format!("failed to write {}", full_md_path.display()))?;
    Ok(Some(full_md_path))
}

fn paddle_markdown_target_rel_path(rel_path: &str, page_index: usize) -> PathBuf {
    let normalized = rel_path.trim().trim_start_matches('/').replace('\\', "/");
    if normalized
        .split_once('/')
        .map(|(prefix, _)| is_page_prefix(prefix))
        .unwrap_or_else(|| is_page_prefix(&normalized))
    {
        return PathBuf::from(normalized);
    }
    PathBuf::from(format!("page-{page_index}")).join(normalized)
}

fn paddle_markdown_target_rel_path_for_image_key(
    rel_path: &str,
    markdown_text: &str,
    page_index: usize,
) -> PathBuf {
    let normalized = rel_path.trim().trim_start_matches('/').replace('\\', "/");
    if normalized
        .split_once('/')
        .map(|(prefix, _)| is_page_prefix(prefix))
        .unwrap_or_else(|| is_page_prefix(&normalized))
    {
        return PathBuf::from(normalized);
    }
    let re = Regex::new(r#"(?i)<img\b[^>]*\bsrc=["']([^"']+)["']"#).expect("valid img src regex");
    for captures in re.captures_iter(markdown_text) {
        let src = captures[1]
            .trim()
            .trim_start_matches('/')
            .replace('\\', "/");
        if src.ends_with(&normalized)
            && src
                .split_once('/')
                .map(|(prefix, _)| is_page_prefix(prefix))
                .unwrap_or_else(|| is_page_prefix(&src))
        {
            return PathBuf::from(src);
        }
    }
    paddle_markdown_target_rel_path(&normalized, page_index)
}

fn paddle_markdown_rel_src_path(src: &str, page_index: usize) -> String {
    let normalized = src.trim().trim_start_matches('/').replace('\\', "/");
    if normalized.is_empty()
        || normalized.starts_with("http://")
        || normalized.starts_with("https://")
        || normalized.starts_with("data:")
    {
        return normalized;
    }
    if normalized.starts_with("images/") {
        return normalized;
    }
    PathBuf::from("images")
        .join(paddle_markdown_target_rel_path(&normalized, page_index))
        .to_string_lossy()
        .replace('\\', "/")
}

fn rewrite_paddle_markdown_image_srcs(text: &str, page_index: usize) -> String {
    let re =
        Regex::new(r#"(?i)(<img\b[^>]*\bsrc=["'])([^"']+)(["'])"#).expect("valid img src regex");
    re.replace_all(text, |captures: &Captures<'_>| {
        format!(
            "{}{}{}",
            &captures[1],
            paddle_markdown_rel_src_path(&captures[2], page_index),
            &captures[3]
        )
    })
    .into_owned()
}

fn normalize_paddle_markdown_images(text: &str, page_index: usize) -> String {
    let rewritten = rewrite_paddle_markdown_image_srcs(text, page_index);
    let centered_div_re = Regex::new(
        r#"(?i)<div\s+style=["']text-align:\s*center;?["']\s*>\s*(<img\b[^>]*>)\s*</div>"#,
    )
    .expect("valid centered img div regex");
    let without_center_div = centered_div_re.replace_all(&rewritten, |captures: &Captures<'_>| {
        markdown_image_from_img_tag(&captures[1])
    });
    let img_re = Regex::new(r#"(?i)<img\b([^>]*)>"#).expect("valid img tag regex");
    img_re
        .replace_all(&without_center_div, |captures: &Captures<'_>| {
            markdown_image_from_attrs(&captures[1])
        })
        .into_owned()
}

fn markdown_image_from_img_tag(img_tag: &str) -> String {
    let img_re = Regex::new(r#"(?i)<img\b([^>]*)>"#).expect("valid img tag regex");
    let Some(captures) = img_re.captures(img_tag) else {
        return img_tag.to_string();
    };
    markdown_image_from_attrs(&captures[1])
}

fn markdown_image_from_attrs(attrs_text: &str) -> String {
    let attr_re = Regex::new(r#"([A-Za-z_:][-A-Za-z0-9_:.]*)\s*=\s*["']([^"']*)["']"#)
        .expect("valid html attr regex");
    let mut src = String::new();
    let mut alt = String::from("Image");
    for captures in attr_re.captures_iter(attrs_text) {
        let key = captures[1].to_ascii_lowercase();
        if key == "src" {
            src = captures[2].trim().to_string();
        } else if key == "alt" {
            let value = captures[2].trim();
            if !value.is_empty() {
                alt = value.to_string();
            }
        }
    }
    if src.is_empty() {
        return format!("<img{attrs_text}>");
    }
    let escaped_alt = alt.replace('[', r"\[").replace(']', r"\]");
    format!("![{escaped_alt}]({src})")
}

fn is_page_prefixed_alias(target_rel_path: &str, source_rel_path: &str) -> bool {
    let target = target_rel_path
        .trim()
        .trim_start_matches('/')
        .replace('\\', "/");
    let source = source_rel_path
        .trim()
        .trim_start_matches('/')
        .replace('\\', "/");
    target != source
        && target.ends_with(&source)
        && target
            .split_once('/')
            .map(|(prefix, _)| is_page_prefix(prefix))
            .unwrap_or_else(|| is_page_prefix(&target))
}

fn is_page_prefix(value: &str) -> bool {
    let Some(number) = value.strip_prefix("page-") else {
        return false;
    };
    !number.is_empty() && number.chars().all(|ch| ch.is_ascii_digit())
}

async fn decode_markdown_image_payload(http: &Client, raw_payload: &Value) -> Result<Vec<u8>> {
    let payload = raw_payload
        .as_str()
        .map(str::trim)
        .filter(|value| !value.is_empty())
        .ok_or_else(|| anyhow!("empty paddle markdown image payload"))?;

    if payload.starts_with("http://") || payload.starts_with("https://") {
        let response = http
            .get(payload)
            .send()
            .await
            .with_context(|| format!("failed to download markdown image {payload}"))?
            .error_for_status()
            .with_context(|| format!("markdown image returned error status: {payload}"))?;
        let bytes = response.bytes().await?;
        return Ok(bytes.to_vec());
    }

    if payload.starts_with("data:") {
        let (_, encoded) = payload
            .split_once(',')
            .ok_or_else(|| anyhow!("invalid data-url markdown image payload"))?;
        return STANDARD
            .decode(encoded)
            .context("failed to decode data-url markdown image payload");
    }

    STANDARD
        .decode(payload)
        .context("failed to decode base64 markdown image payload")
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[tokio::test]
    async fn materialize_paddle_markdown_artifacts_writes_full_md_and_images() {
        let root = std::env::temp_dir().join(format!("rust-api-paddle-md-{}", fastrand::u64(..)));
        let payload = json!({
            "layoutParsingResults": [
                {
                    "markdown": {
                        "text": "<div style=\"text-align: center;\"><img src=\"imgs/a.png\" alt=\"Image\" width=\"48%\" /></div>",
                        "images": {
                            "imgs/a.png": "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mP8/x8AAwMCAO+aD3sAAAAASUVORK5CYII="
                        }
                    }
                }
            ]
        });

        let full_md = materialize_paddle_markdown_artifacts(&payload, &root)
            .await
            .expect("materialize")
            .expect("markdown path");
        let content = tokio::fs::read_to_string(&full_md)
            .await
            .expect("read markdown");
        assert!(content.contains("![Image](images/page-1/imgs/a.png)"));
        assert!(!content.contains("<img"));
        assert!(root.join("md/images/page-1/imgs/a.png").exists());

        let _ = std::fs::remove_dir_all(root);
    }

    #[tokio::test]
    async fn materialize_paddle_markdown_artifacts_rewrites_page_prefixed_src_with_unprefixed_key()
    {
        let root = std::env::temp_dir().join(format!("rust-api-paddle-md-{}", fastrand::u64(..)));
        let payload = json!({
            "layoutParsingResults": [
                {
                    "markdown": {
                        "text": "<div><img src=\"page-5/imgs/a.png\" alt=\"Image\" /></div>",
                        "images": {
                            "imgs/a.png": "iVBORw0KGgoAAAANSUhEUgAAAAEAAAABCAQAAAC1HAwCAAAAC0lEQVR42mP8/x8AAwMCAO+aD3sAAAAASUVORK5CYII="
                        }
                    }
                }
            ]
        });

        let full_md = materialize_paddle_markdown_artifacts(&payload, &root)
            .await
            .expect("materialize")
            .expect("markdown path");
        let content = tokio::fs::read_to_string(&full_md)
            .await
            .expect("read markdown");
        assert!(content.contains("![Image](images/page-5/imgs/a.png)"));
        assert!(!content.contains("<img"));
        assert!(root.join("md/images/page-5/imgs/a.png").exists());

        let _ = std::fs::remove_dir_all(root);
    }
}
